x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/FrequentItemSetCollector.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.frequentitemsets;
+
+import org.apache.lucene.util.LongsRef;
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.search.aggregations.Aggregation.CommonFields;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * Collector for frequent item sets.
+ *
+ * The collector keeps the best top-n sets found during mining and reports the minimum doc count required to enter.
+ * With other words the last of top-n. This is useful to skip over candidates quickly.
+ *
+ * Technically this is implemented as priority queue which is implemented as heap with the last item on top.
+ *
+ * To get to the top-n results pop the collector until it is empty and than reverse the order.
+ */
+public class FrequentItemSetCollector {
+
+    /**
+     * Container for a single frequent itemset
+     */
+    public class FrequentItemSet implements ToXContent {
+
+        private LongsRef items;
+        private long docCount;
+
+        // every set has a unique id, required for the outer logic
+        private int id;
+
+        private FrequentItemSet() {
+            this.id = -1;
+            this.items = new LongsRef(10);
+            this.docCount = -1;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            final long totalTransactionCount = transactionStore.getTotalTransactionCount();
+
+            builder.startObject();
+            builder.startObject(CommonFields.KEY.getPreferredName());
+            Map<String, List<String>> frequentItemsKeyValues = new HashMap<>();
```

Comment:
How would a keyed response look like?

I can only think of:

```
"key" : {[
   {
     field: "a",
     value: [1,2,3]
   },
   {
     field: "b",
     value: [4,5,6]
   }
]}
```

I will add this to the design doc, however I think `keyed:false` is a nicetohave.

The line you marked has a different purpose: A frequent item set is a list of items, e.g. `a:1`, `b:4`, `c:66`, `a:2`, ...
What this does it grouping the `a`'s together in order to return `a:1,2, ...`.  For sure this can be done smarter, so this has room for improvement.